packages/adblocker-electron/src/index.ts

```diff
@@ -164,101 +166,80 @@ export class ElectronBlocker extends FiltersEngine {
-      getInjectionRules: true,
-      getExtendedRules: true,
-      getRulesFromHostname: true,
-      getRulesFromDOM: false, // Only done on updates (see `onGetCosmeticFiltersUpdated`)
-
-      callerContext: {
-        frameId: event.frameId,
-        processId: event.processId,
-      },
-    });
-
-    if (active === false) {
-      event.returnValue = null;
-      return;
-    }
-
-    // Inject custom stylesheets
-    this.injectStyles(event.sender, styles);
-
-    event.sender.send('get-cosmetic-filters-response', {
-      active,
-      extended,
-      styles: '',
-    } as IMessageFromBackground);
-
-    // to execute Inject scripts synchronously, simply return scripts to renderer.
-    event.returnValue = scripts;
+  public onIsMutationObserverEnabled = async (
+    _: Electron.IpcMainInvokeEvent,
+  ): Promise<boolean> => {
+    return this.config.enableMutationObserver;
   };
 
-  public onGetCosmeticFiltersUpdated = (
-    event: Electron.IpcMainEvent,
+  public onInjectCosmeticFilters = async (
+    event: Electron.IpcMainInvokeEvent,
     url: string,
-    msg: IBackgroundCallback,
-  ): void => {
-    // Extract hostname from sender's URL
+    msg?: IBackgroundCallback,
+  ): Promise<CosmeticFiltersResponse> => {
     const parsed = parse(url);
     const hostname = parsed.hostname || '';
     const domain = parsed.domain || '';
 
-    const { active, styles, extended } = this.getCosmeticsFilters({
+    // `msg` is undefined for the initial call and present for subsequent updates
+    const isFirstRun = msg === undefined;
+
+    const { active, styles, scripts } = this.getCosmeticsFilters({
       domain,
       hostname,
       url,
 
-      classes: msg.classes,
-      hrefs: msg.hrefs,
-      ids: msg.ids,
+      // DOM information, only available for updates
+      classes: msg?.classes,
+      hrefs: msg?.hrefs,
+      ids: msg?.ids,
 
-      // Only done on first load in the frame, disable for updates
-      getBaseRules: false,
-      getInjectionRules: false,
+      // Rules to fetch: true for initial call, false for updates
+      getBaseRules: isFirstRun,
+      getInjectionRules: isFirstRun,
       getExtendedRules: false,
-      getRulesFromHostname: false,
+      getRulesFromHostname: isFirstRun,
 
-      // This will be done every time we get information about DOM mutation
-      getRulesFromDOM: true,
+      // Only true for update calls when we have DOM information
+      getRulesFromDOM: !isFirstRun,
 
       callerContext: {
         frameId: event.frameId,
         processId: event.processId,
-
-        lifecycle: msg.lifecycle,
+        lifecycle: msg?.lifecycle,
       },
     });
 
     if (active === false) {
-      return;
+      return { active: false };
     }
 
-    // Inject custom stylesheets
-    this.injectStyles(event.sender, styles);
+    if (styles.length > 0) {
+      try {
+        await event.sender.insertCSS(styles, { cssOrigin: 'user' });
+      } catch (error) {
+        return {
```

Comment:
this is a problem - if you fail to inject CSS, you don't try to inject scriptlets

Reply:
I'll address this, thank you for catching it.

Reply:
Seems like we don't need to await on insertCSS, or handle errors, we simply call it and proceed to script injection right now. Please let me know if we should do it in order.